ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+    -------
+    wrapper : Callable
+        The wrapped function
+    """
+    if parameter_names is None:
+        parameter_names = ['v{}'.format(i) for i in range(len(input_types))]
+    else:
+        assert len(input_types) == len(parameter_names)
+
+    udf_node_fields = collections.OrderedDict([
+        (name, Arg(rlz.value(type_)))
+        for name, type_ in zip(parameter_names, input_types)
+    ] + [
+        (
+            'output_type',
+            lambda self, output_type=output_type: rlz.shape_like(
+                self.args, dtype=output_type
+            )
+        )
+    ])
+    udf_node_fields['resolve_name'] = lambda self: name
+
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sqlalchemy.func
+        if schema is not None:
+            func_obj = getattr(func_obj, schema)
+        func_obj = getattr(func_obj, name)
+
+        sa_args = [t.translate(arg) for arg in expr.op().args]
+
+        return func_obj(*sa_args)
+
+    add_operation(udf_node, _translate_udf)
+
+    def wrapped(*args, **kwargs):
+        node = udf_node(*args, **kwargs)
+        return node.to_expr()
+
+    return wrapped
+
+
+def func_to_udf(conn,
+                python_func,
+                in_types=None,
+                out_type=None,
+                schema=None,
+                name=None,
+                overwrite=False):
+    """Defines a UDF in the database
+
+    Parameters
+    ----------
+    conn: sqlalchemy engine
+    python_func: python function
+    in_types: List[DataType]; if left None, will try to infer datatypes from
+    function signature
+    out_type : DataType
+    schema: str - optionally specify the schema in which to define the UDF
+    name: str - name for the UDF to be defined in database
+    overwrite: bool - replace UDF in database if already exists
+
+    Returns
+    -------
+    wrapper : Callable
+        The ibis UDF object as a wrapped function
+    """
+    if name is None:
+        internal_name = python_func.__name__
+    else:
+        internal_name = name
+    signature = inspect.signature(python_func)
+    parameter_names = signature.parameters.keys()
+    if in_types is None:
+        raise NotImplementedError('inferring in_types not implemented')
+    if out_type is None:
+        raise NotImplementedError('inferring out_type not implemented')
+    replace = ' OR REPLACE ' if overwrite else ''
+    schema_fragment = (schema + '.') if schema else ''
+    template = """CREATE {replace} FUNCTION
+{schema_fragment}{name}({signature})
+RETURNS {return_type}
+LANGUAGE plpythonu
+AS $$
+{func_definition}
+return {internal_name}({args})
+$$;
+"""
+
+    postgres_signature = ', '.join(
+        '{name} {type}'.format(
+            name=name,
+            type=get_sqltype(type_),
+        )
+        for name, type_ in zip(parameter_names, in_types)
+    )
+    return_type = get_sqltype(out_type)
+    func_definition = inspect.getsource(python_func)
```

Comment:
Something to be aware of is that `inspect.getsource` only works if `python_func` is defined in a file. This means that in a vanilla `python` session (*not* IPython) you won't be able to define a UDF using this mechanism.

This isn't a showstopper IMO; the utility of this feature is high enough that it's not a hard requirement to support every possible scenario that folks may use it in.

Reply:
I played around with a scrap script/module, and `inspect.getsource` **_is_** still able to get the source, even when in vanilla python and not in IPython. 

That said, there are certainly still edge cases that my current implementation doesn't cover. For instance, it won't work if the function uses any variables defined outside the scope of the function itself or any modules imported outside the function. There could be ways of automatically detecting imports that need to be inserted or sucking in globals, but I would propose that **for now it can be the user's responsibility to make sure their function is entirely self-contained**.